docs/design/2018-09-04-histograms-in-plan.md

```diff
@@ -0,0 +1,115 @@
+
+For `Join`, there’re joins as follows:
+
+- Inner join: use histograms to do the row count estimation with the join key condition. Since it won’t have one side filter, we only need to consider the composite filters after considering the join key. We can simply multiply `selectionFactor` if there are other composite filters in our first version of implementation. Since `Selectivity` cannot calculate selectivity of expression that containing multiple column.
+
+- One side outer join: It depends on the join keys’ NDV. And we can just use histograms to estimate it if there’re non-join-key filters.
+
+- Semi join: It’s something similar to inner join. But no data expanding occurs. When we maintain the range information. We can get a nearly accurate answer of its row count.
+
+- Anti semi join: Same with semi join.
+
+For `Selection`, just use it to calculate the selectivity. 
+
+For `DataSource`, if it’s a non-partitioned table, we just maintain the map. If it’s a partitioned table, we now only store the statistics of each partition So we need to merge them. We’ll need a cache or something else to ensure that we won’t merge them each time we need it, which will consume tooooo much time and memory space.
+
+For other plan operators or `DataSource` which do not contain any histogram, we just use the original way to maintain `statsInfo`. We won’t maintain histograms for them. Taking `Union` for an example, if we maintain histogram for it, we need to merge the histograms which is really an expensive operation. But we can consider maintaining this if a hint is given.
+
+When we maintain `rangesOfXXX` or `max and min values`:
+
+Most things are the same with above. But there will be something more to do.
+
+For `Selection` and `Join`, we need to cut off the things which are not in ranges when doing estimation. And update the ranges information after the estimation. Also the NDV of column can be estimated more accurately.
+
+For `Aggregation`, we only need to cut off the things which are not in ranges when doing estimation. There is no need to update the ranges information.
+
+For `TopN`, we now have alibity to maintain histograms of the order-by items.
+
+
+## Rationale
+
+### How other systems solve the same issue?
+
+I’ve looked into Spark. They did nearly the same thing with what I said. They only maintain the max and min values, no `ranges` information. And they don’t have index, so they only maintain the column’s max/min value which make problem far more much easier.
+
+As for Orca and Calcite, I haven’t discovered where they maintain this information. But there’s something about statistics in Orca’s paper. According to the paper, I think they construct new histogram during planning and cache it for not building to often.
```

Comment:
... I think they construct **a** new histogram during planning and cache it **to avoid building too many times**.
or 
... I think they construct **a** new histogram during planning and cache it **to avoid repeated building**.